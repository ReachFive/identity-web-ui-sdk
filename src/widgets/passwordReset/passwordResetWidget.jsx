import React from 'react';

import pick from 'lodash-es/pick';
import omit from 'lodash-es/omit';

import { parseQueryString } from '../../helpers/queryString'

import { Heading, Info, Link } from '../../components/miscComponent';
import { createMultiViewWidget } from '../../components/widget/widget';
import { PasswordEditorForm } from '../passwordEditor/passwordEditorWidget'

class MainView extends React.Component {
    static defaultProps = {
        onSuccess: () => { },
        onError: () => { }
    };

    handleSubmit = data => {
        return this.props.apiClient.updatePassword({
            ...omit(data, ['passwordConfirmation']),
            ...this.props.authentication
        });
    };

    onSuccess = () => {
        this.props.onSuccess();
        this.props.goTo('success');
    };

    render() {
        const { i18n } = this.props;

        return <div>
            <Heading>{i18n('passwordReset.title')}</Heading>
            <Info>{i18n('passwordReset.intro')}</Info>
            <PasswordEditorForm
                handler={this.handleSubmit}
                canShowPassword={this.props.canShowPassword}
                showLabels={this.props.showLabels}
                onSuccess={this.onSuccess}
                onError={this.props.onError} />
        </div>;
    }
}

const SuccessView = ({ i18n, loginLink }) => (
    <div>
        <Heading>{i18n('passwordReset.successTitle')}</Heading>
        <Info>{i18n('passwordReset.successMessage')}</Info>
        {loginLink && (
            <Info>
                <Link href={loginLink}>{i18n('passwordReset.loginLink')}</Link>
            </Info>
        )}
    </div>
);

const resolveCode = () => {
    const qs = (window.location.search && window.location.search.length)
        ? window.location.search.substr(1)
        : '';

    return { authentication: pick(parseQueryString(qs), ['verificationCode', 'email', 'purpose']) };
};

export default createMultiViewWidget({
    initialView: 'main',
    views: {
        main: MainView,
        success: SuccessView
    },
    prepare: options => ({
        ...options,
        ...resolveCode()
    })
});
