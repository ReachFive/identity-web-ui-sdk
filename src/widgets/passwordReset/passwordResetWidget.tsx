import React from 'react';

import { parseQueryString } from '../../helpers/queryString'

import { Heading, Info, Link } from '../../components/miscComponent';
import { createMultiViewWidget } from '../../components/widget/widget';
import { PasswordEditorForm, PasswordEditorFormData } from '../passwordEditor/passwordEditorWidget'
import { useReachfive } from '../../contexts/reachfive';
import { useRouting } from '../../contexts/routing';
import { useI18n } from '../../contexts/i18n';

interface MainViewProps {
    /**
     * Whether or not to provide the display password in clear text option.
     * @default false
     */
    canShowPassword?: boolean
    /**
     * Callback function called when the request has failed.
     */
    onSuccess?: () => void
    /**
     * Callback function called after the widget has been successfully loaded and rendered inside the container.
     * The callback is called with the widget instance.
     */
    onError?: () => void
    /**
     * Whether the form fields' labels are displayed on the form view.
     * @default false
     */
    showLabels?: boolean
}

const MainView = ({
    authentication,
    canShowPassword = false,
    onSuccess = () => {},
    onError = () => {},
    showLabels = false,
}: PropsWithAuthentication<MainViewProps>) => {
    const coreClient = useReachfive()
    const i18n = useI18n()
    const { goTo } = useRouting()

    const handleSubmit = ({ password }: PasswordEditorFormData) => {
        return coreClient.updatePassword({
            password,
            ...authentication
        });
    };

    const handleSuccess = () => {
        onSuccess();
        goTo('success');
    };

    return (
        <div>
            <Heading>{i18n('passwordReset.title')}</Heading>
            <Info>{i18n('passwordReset.intro')}</Info>
            <PasswordEditorForm
                handler={handleSubmit}
                canShowPassword={canShowPassword}
                showLabels={showLabels}
                onSuccess={handleSuccess}
                onError={onError} />
        </div>
    )
}

interface SuccessViewProps {
    loginLink?: string
}

const SuccessView = ({ loginLink }: SuccessViewProps) => {
    const i18n = useI18n()
    return  (
        <div>
            <Heading>{i18n('passwordReset.successTitle')}</Heading>
            <Info>{i18n('passwordReset.successMessage')}</Info>
            {loginLink && (
                <Info>
                    <Link href={loginLink}>{i18n('passwordReset.loginLink')}</Link>
                </Info>
            )}
        </div>
    )
}

const resolveCode = () => {
    const qs = window.location.search.substring(1)
    const { verificationCode, email } = parseQueryString(qs)
    return { authentication: { verificationCode, email } as Authentication };
};

type Authentication = { verificationCode: string,  email: string }
type PropsWithAuthentication<P> = P & { authentication?: Authentication }

export interface PasswordResetWidgetProps extends MainViewProps, SuccessViewProps {}

export default createMultiViewWidget<PasswordResetWidgetProps, PropsWithAuthentication<PasswordResetWidgetProps>>({
    initialView: 'main',
    views: {
        main: MainView,
        success: SuccessView
    },
    prepare: options => ({
        ...options,
        ...resolveCode()
    })
});
