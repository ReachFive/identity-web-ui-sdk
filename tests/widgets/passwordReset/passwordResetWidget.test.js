import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { render } from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import passwordResetWidget from '../../../src/widgets/passwordReset/passwordResetWidget'

Enzyme.configure({ adapter: new Adapter() });

describe('Snapshot', () => {
    const generateSnapshot = (options) => () => {
        const tree = passwordResetWidget(options, { config: {}, apiClient: {} })
            .then(result => renderer.create(result).toJSON());

        expect(tree).resolves.toMatchSnapshot();
    };

    describe('password-reset', () => {
        test('default', generateSnapshot());
    });
});

describe('DOM testing', () => {
    const generateComponent = async (options) => {
        const result = await passwordResetWidget(options, { config: {}, apiClient: {} });

        return render(result);
    };

    describe('password-reset', () => {
        test('basic', async () => {
            expect.assertions(3);

            // When
            const instance = await generateComponent();

            // Then
            expect(instance.find('input[name="password"]')).toHaveLength(1);
            expect(instance.find('input[name="password_confirmation"]')).toHaveLength(1);
            expect(instance.find('button').text()).toBe('Send');
        });
    });
});
