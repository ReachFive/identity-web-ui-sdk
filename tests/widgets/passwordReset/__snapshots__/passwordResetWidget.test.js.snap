// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot password-reset default 1`] = `
.c1 {
  margin-bottom: 15px;
  text-align: center;
  color: #212529;
  font-weight: bold;
  font-size: 16.8px;
}

.c2 {
  text-align: center;
  margin-bottom: 10px;
  color: #495057;
}

.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
  padding: 20px;
  border-radius: 3px;
  background-color: #ffffff;
  max-width: 400px;
  box-sizing: border-box;
  margin: 0 auto;
}

.c7 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  font-weight: bold;
  white-space: nowrap;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  cursor: pointer;
  color: #ffffff;
  background-color: #229955;
  border: 1px solid #229955;
  padding: 9px 12px;
  font-size: 14px;
  line-height: 1.428571429;
  border-radius: 3px;
  -webkit-transition: all .15s ease-in-out;
  transition: all .15s ease-in-out;
}

.c7:focus {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(34,153,85,0.5);
}

.c7:hover,
.c7:active {
  background-color: #1b7842;
  border-color: #1b7842;
}

.c7[disabled] {
  opacity: .65;
}

.c3 {
  position: relative;
}

.c5 {
  color: #495057;
  margin-bottom: 5px;
  display: none;
}

.c4 {
  margin-bottom: 10px;
}

.c6 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.428571429;
  color: #495057;
  border-radius: 3px;
  border-width: 1px;
  border-style: solid;
  border-color: #ced4da;
  background-color: #fff;
  background-image: none;
  padding: 9px 12px;
  -webkit-transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  box-shadow: none;
  -webkit-appearance: none;
}

.c6:focus {
  border-color: #5fdb94;
  box-shadow: 0 0 0 3px rgba(95,219,148,0.5);
  outline: 0;
}

.c6:disabled,
.c6[readonly] {
  background-color: #e9ecef;
}

.c6::-webkit-input-placeholder {
  color: #868e96;
}

.c6::-moz-placeholder {
  color: #868e96;
}

.c6:-ms-input-placeholder {
  color: #868e96;
}

.c6::placeholder {
  color: #868e96;
}

<div
  className="c0"
>
  <div>
    <div
      className="c1"
    >
      Reset your password
    </div>
    <div
      className="c2"
    >
      Please type your new password.
    </div>
    <form
      className="c3"
      noValidate={true}
      onSubmit={[Function]}
    >
      <div
        className="c4"
      >
        <div
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <label
            className="c5"
          >
            New password
          </label>
          <input
            autoComplete="new-password"
            className="c6"
            data-testid="password"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="New password"
            title="New password"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        className="c4"
      >
        <label
          className="c5"
          htmlFor="password_confirmation"
        >
          Password confirmation
        </label>
        <div
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <input
            autoComplete="new-password"
            className="c6"
            data-testid="password_confirmation"
            id="password_confirmation"
            name="password_confirmation"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Password confirmation"
            required={true}
            title="Password confirmation"
            type="password"
            value=""
          />
        </div>
      </div>
      <button
        className="c7"
        data-testid="submit"
        disabled={false}
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
</div>
`;
